### ISSUE TITLE
DAT - Potential difference between the paper & the code

### ISSUE THREAD

**Issue by ksachdeva** [mmalekzadeh/dana#3]
Hi,

I am using your 13th august copy of the paper.

The algorithm DAT from the paper is:

![image](https://user-images.githubusercontent.com/127006/130832671-2b7c28e4-5265-4589-afd9-ff7df5a0bf0e.png)

When I read this algorithm, I get the impression - 

For every iteration in an epoch, we would construct few batches e.g. B=5 batches

a) Compute the loss for the given batch (in which dimension randomization has been applied)
b) Compute the gradient for this batch
c) Accumulate the gradients. Most importantly - do not apply the gradients

Once we have run the batches (i.e. 5 of them) then apply the accumulated gradients.

Now, when I look at the code, I see the following:

```
for epoch in range(num_epochs):  

        ## Training
        train_dataset = tf.data.Dataset.from_tensor_slices((X_train, Y_train))
        train_dataset = iter(train_dataset.shuffle(len(X_train)).batch(batch_size))
        n_iterations_per_epoch = len(X_train)//(batch_size*n_batch_per_train_setp)
        epoch_loss_avg = tf.keras.metrics.Mean()           

        for i in range(n_iterations_per_epoch):
            rnd_order_H = np.random.permutation(len(H_combinations))
            rnd_order_W = np.random.permutation(len(W_combinations))
            n_samples = 0.
            with tf.GradientTape() as tape:
                accum_loss = tf.Variable(0.)
                for j in range(n_batch_per_train_setp):
                    try:
                        X, Y = next(train_dataset)
                    except:
                        break
                    X = X.numpy()
                    sample_weight = [data_class_weights[y] for y in Y.numpy()]
                    
                    ### Dimension Randomization 
                    ####### Random Sensor Selection
                    rnd_H = H_combinations[rnd_order_H[j%len(rnd_order_H)]]                    
                    X = X[:,:,rnd_H,:] 
                    ####### Random Sampling Rate Selection  
                    rnd_W = W_combinations[rnd_order_W[j%len(rnd_order_W)]]
                    X = tf.image.resize(X, (rnd_W, len(rnd_H)))    

                    logits =  model(X)               
                    accum_loss = accum_loss + loss_fn(Y, logits, sample_weight)
                    n_samples = n_samples + 1.
            gradients = tape.gradient(accum_loss, model.trainable_weights)
            gradients = [g*(1./n_samples) for g in gradients]
            optimizer.apply_gradients(zip(gradients, model.trainable_weights))
            epoch_loss_avg.update_state(accum_loss*(1./n_samples))
```

If I understood the code properly, you 
- accumulate the loss for the 5 batches. 
- then compute the gradient on this accumulated loss.
- then take the average of gradients
- then apply the gradients

The flow in code vs the algorithm in the paper seems different but maybe the end result is the same.

Would appreciate it if you could clarify/confirm.

Regards & thanks
Kapil

**Comment by mmalekzadeh** [comment]
Hi @ksachdeva 

Good point: both have the same result.

When you call loss function, it computes the gradient of the current batch and accumulates the computed gradients in the loop. Basically, this implementation "makes it simple to take the gradient of the sum of a collection of losses".

For reference, please look at this: 
https://www.tensorflow.org/guide/autodiff#gradients_of_non-scalar_targets

Best
Mohammad

**Comment by mmalekzadeh** [comment]
Hey @ksachdeva 
Please let me know if this is still an issue. Thanks
